### ISSUE TITLE
Use traits and extension traits for explicit request methods

### ISSUE THREAD

**Issue by notgull** [psychon/x11rb#694]
Right now, the `generator` crate generates two sets of protocol: one in the `x11rb-protocol` crate and one in the `x11rb` crate. The greater majority of the useful logic is in `x11rb-protocol`, and the parts in `x11rb` are mostly convenience functions. It would be nice if we could move the generated code that's currently in `x11rb` to `x11rb_protocol`, if not just to reduce the complexity of the generator, and to reduce the overall amount of generated code.

My proposal is to have a `SendTarget` trait, like this:

```rust
trait SendTarget<Req> where Req: ReplyRequest {
    type Output;

    fn send(&mut self, req: Req) -> Output;
}
```

Since there are three types of request traits with different expected outputs (`VoidRequest`, `ReplyRequest` and `ReplyRequestWithFds`), we would probably use three `SendTarget` traits to avoid overlap (since there's no real way to tell the compiler that types that implement `VoidRequest` will never implement `ReplyRequest` and `ReplyRequestWithFds`, and so on and so forth).

Then, we'd implement them on `x11rb::RequestConnection` like so:

```rust
impl<'c, Conn: RequestConnection + ?Sized, Req: ReplyRequest> SendTarget<Req> for &'c Conn {
    type Output = Result<Cookie<'c, Conn, Req::Reply>, ConnectionError>;

    fn send(&mut self, req: Req) -> Self::Output {
        /* ... */
    }
}
```

Then, you could implement the utility functions like this:

```rust
pub fn foo<S: SendTarget<FooRequest>>(conn: &mut S, foo: u32) -> S::Output {
    let req = FooRequest { foo };
    conn.send(req);
}
```

Finally, you could also reimplement utility traits (`ConnectionExt`) to use `SendTarget` instead of `RequestConnection`.

What are your thoughts on this @psychon?

**Comment by psychon** [comment]
Quick first reaction: The current convenience code for `PutImage` can use an `IoSlice` and does not have to copy all the pixel data for sending (well, perhaps copy it into the output buffer, but if it is too large for that, then it might still be send without copy). The current trait-based request sending has to put the whole request into a single `Vec<u8>`, because of lifetime issues. This is why I left the code in its current "there are two ways to send requests"-state. I think I already commented on this on some other issue that you opened, but I don't have much time to look for that currently:

TL;DR: Yup, the goal sounds great, but I am worried about performance regressions due to this design (even though I always say that performance goals should be measured and not just assumed, I just assumed...)

**Comment by notgull** [comment]
Yeah. I’ll whip up a branch this week, and see how the `PutImage` request suffers.

**Comment by notgull** [comment]
Just experimented with this; I don't think it works. It relies on being able to implement the foreign trait (e.g. `SendTarget`) on a local generic type (e.g. `Conn: RequestConnection`), which Rust's trait resolver can't do.
